src/torchcodec/samplers/_implem.py

```diff
@@ -323,3 +331,236 @@ def clips_at_regular_indices(
+        )
+
+    all_clips_timestamps_sorted, argsort = zip(
+        *sorted(
+            (frame_index, i) for (i, frame_index) in enumerate(all_clips_timestamps)
+        )
+    )
+    previous_decoded_frame = None
+    all_decoded_frames = [None] * len(all_clips_timestamps)
+    for i, j in enumerate(argsort):
+        frame_pts_seconds = all_clips_timestamps_sorted[i]
+        if (
+            previous_decoded_frame is not None  # then we know i > 0
+            and frame_pts_seconds == all_clips_timestamps_sorted[i - 1]
+        ):
+            # Avoid decoding the same frame twice.
+            # IMPORTANT: this is only correct because a copy of the frame will
+            # happen within `to_framebatch` when we call torch.stack.
+            # If a copy isn't made, the same underlying memory will be used for
+            # the 2 consecutive frames. When we re-write this, we should make
+            # sure to explicitly copy the data.
+            decoded_frame = previous_decoded_frame
+        else:
+            decoded_frame = decoder.get_frame_displayed_at(seconds=frame_pts_seconds)
+        previous_decoded_frame = decoded_frame
+        all_decoded_frames[j] = decoded_frame
+
+    all_clips: list[list[Frame]] = chunk_list(
+        all_decoded_frames, chunk_size=num_frames_per_clip
+    )
+
+    return [to_framebatch(clip) for clip in all_clips]
+
+
+def clips_at_regular_timestamps(
+    decoder,
+    *,
+    seconds_between_clip_starts: float,
+    num_frames_per_clip: int = 1,
+    seconds_between_frames: Optional[float] = None,
+    # None means "begining", which may not always be 0
+    sampling_range_start: Optional[float] = None,
+    sampling_range_end: Optional[float] = None,  # interval is [start, end).
+    policy: str = "repeat_last",
+) -> List[FrameBatch]:
+    # Note: *everywhere*, sampling_range_end denotes the upper bound of where a
+    # clip can start. This is an *open* upper bound, i.e. we will make sure no
+    # clip starts exactly at (or above) sampling_range_end.
+
+    _validate_params(
+        decoder=decoder,
+        num_frames_per_clip=num_frames_per_clip,
+        policy=policy,
+    )
+
+    seconds_between_frames = _validate_params_time_based(
+        decoder=decoder,
+        seconds_between_clip_starts=seconds_between_clip_starts,
+        seconds_between_frames=seconds_between_frames,
+    )
+
+    sampling_range_start, sampling_range_end = _validate_sampling_range_time_based(
+        num_frames_per_clip=num_frames_per_clip,
+        seconds_between_frames=seconds_between_frames,
+        sampling_range_start=sampling_range_start,
+        sampling_range_end=sampling_range_end,
+        begin_stream_seconds=decoder.metadata.begin_stream_seconds,
+        end_stream_seconds=decoder.metadata.end_stream_seconds,
+    )
+
+    clip_start_seconds = torch.arange(
+        sampling_range_start,
+        sampling_range_end,  # excluded
+        seconds_between_clip_starts,
+    )
```

Comment:
I'm fairly confident that implementing a random time-based sampler should be trivial, i.e. just changing this to

```py
clip_start_seconds = torch.rand(num_clips) * (sampling_range_end - sampling_range_start) + sampling_range_start
```

LMK if we want this and I'll follow-up with a PR.